### PR TITLE
std.format: Replace snprintf for x87-reals

### PR DESCRIPTION
**Pull request by berni44**
Finally going life with formatting x87-reals.

I consider this a fix for several issues, although the issues might still remain for some exotic reals: There is no official list, which reals D does support and also no way for me to test if the issues still exist. That means, the issues can never be closed if we do not decide at some time, that it's enough and anyway any fix of these issues will high likely be to support printing floats for that exotic reals, so there is no benefit in leaving the issues open.

### CODE REVIEW

**Comment by thewilsonator** [comment]
> There is no official list, which reals D does support and also no way for me to test if the issues still exist. 

From memory this is:
* IEEE 32 bit (float),
* IEEE 64 bit (double), 
* intel 80-bit x87 ,
* IBM double-double, 
* IEEE 128-bit

Take a look at `RealFormat` in phobos.

**Comment by thewilsonator** [comment]
> so there is no benefit in leaving the issues open.

Sounds good to me.

**Review by thewilsonator — APPROVED** [review]

**Comment by berni44** [comment]
> > There is no official list, which reals D does support and also no way for me to test if the issues still exist.
> 
> From memory this is:
> 
> * IEEE 32 bit (float),
> * IEEE 64 bit (double),
> * intel 80-bit x87 ,
> * IBM double-double,
> * IEEE 128-bit
> 
> Take a look at `RealFormat` in phobos.

`RealFormat` (which is not public) defines:

```
    ieeeHalf,
    ieeeSingle,
    ieeeDouble,
    ieeeExtended,   // x87 80-bit real
    ieeeExtended53, // x87 real rounded to precision of double.
    ibmExtended,    // IBM 128-bit extended
    ieeeQuadruple,
```

Docs of `std.math` state:

- 64 bit Big-endian  'double' (eg PowerPC))
- 128 bit Big-endian 'quadruple' (eg SPARC))
- 64 bit Little-endian 'double' (eg x86-SSE2))
- 80 bit Little-endian, with implied bit 'real80' (eg x87, Itanium))
- 128 bit Little-endian 'quadruple' (not implemented on any known processor!))
- Non-IEEE 128 bit Big-endian 'doubledouble' (eg PowerPC) has partial support)

Language reference doesn't state anything, as far as I know.

File: changelog/formatting_floats_in_CTFE.dd [hunk 1]
```diff
@@ -1,4 +1,4 @@
-`float` and `double` values can be formatted at compile time
+`float`, `double` and certain `real` values can be formatted at compile time
```
Comment (PetarKirov) [inline changelog/formatting_floats_in_CTFE.dd:1]: 64-bit and 80-bit `real`s accounts for 100% of the platforms supported by dmd, so I think the note below is sufficient.

```suggestion
`float`, `double` and `real` values can be formatted at compile time
```

File: std/format/internal/write.d [hunk 2]
```diff
@@ -1031,6 +1033,30 @@ if (is(FloatingPointTypeOf!T) && !is(T == enum) && !hasToString!(T, Char))
     assert(format("%#.6g",a) == "-1.00000e+06");
 }
 
+// https://issues.dlang.org/show_bug.cgi?id=8424
+@safe pure unittest
+{
+    static assert(format("%s",0.6f) == "0.6");
+    static assert(format("%s",0.6) == "0.6");
+    static assert(format("%s",0.6L) == "0.6");
```
Comment (PetarKirov) [inline std/format/internal/write.d:1041]:
```suggestion
    static assert(format("%s", 0.6f) == "0.6");
    static assert(format("%s", 0.6) == "0.6");
    static assert(format("%s", 0.6L) == "0.6");
```

File: std/format/internal/write.d [hunk 3]
```diff
@@ -648,7 +649,7 @@ if (is(FloatingPointTypeOf!T) && !is(T == enum) && !hasToString!(T, Char))
             return;
         }
 
-        enforceFmt(!__ctfe, "Cannot format reals at compile-time.");
+        enforceFmt(!__ctfe, "The reals on this computer are not supported at compile-time.");
```
Comment (PetarKirov) [inline std/format/internal/write.d:652]:
```suggestion
        enforceFmt(!__ctfe, mixin(`"Unsupported `real` type: real.sizeof = `, real.sizeof, ` | real.mant_dig = `, real.mant_dig, `"`));
```

**Review by PetarKirov — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]
Awesome! Thanks!

File: changelog/formatting_floats_in_CTFE.dd (hunk 1 above)
Comment (berni44) [reply changelog/formatting_floats_in_CTFE.dd:1]: I suggest "Floating point numbers can be ..."? No need to list all of them...

**Pull request merged**